feat: initial implementation

sitepark-veltrup opened the pull request.

sitepark-schaeper commented on composer.json:
```diff
@@ -0,0 +1,87 @@
+                "version": "dev-feature/initial-implementation"
+            }
+        }
+    }
```
This is the result of `composer link` and only relevant for the development, right?

sitepark-veltrup replied:
is removed

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/PersonDocumentEnricher.php:
```diff
@@ -0,0 +1,168 @@
+
+        $doc->sp_citygov_firstname = $firstname;
+        $doc->sp_citygov_lastname = $lastname;
+        $doc->sp_sortvalue = $lastname . 'aaa' . $firstname;
```
I don't understand what we're doing here, would you mind explaining it to me?

sitepark-veltrup replied:
comment added

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/PersonDocumentEnricher.php:
```diff
@@ -0,0 +1,168 @@
+                continue;
+            }
+
+            $organisationResource = $this->resourceLoader->load(
+                $organisationLocation
+            );
```
What if this fails?

sitepark-veltrup replied:
Reacting to errors in the enricher is not helpful, as the document is always indexed incompletely. All errors are intercepted by the indexer per document. If an error occurs for a document, the error is logged and the document is not indexed.

sitepark-schaeper replied:
Then I believe it would have to declare that/what it can throw. Idealy thats a `DocumentEnrichingException` of some sorts so an implementing source can catch specifically those from any `DocumentEnricher` without having to know what implementations are or could be configured and what those may throw.

sitepark-veltrup replied:
Is that what you mean?
```php
    /**
     * @throws DocumentEnrichingException
     */
    public function enrichDocument(
        Resource $resource,
        IndexDocument $doc,
        string $processId
    ): IndexDocument {
        ...
    }
```

Where the enricher is called, however, I would catch all errors in any case, as I would not rely on all enricher catching all error cases.

```php
            try {
                $doc = $update->createDocument();
                foreach ($this->documentEnricherList as $enricher) {
                    $doc = $enricher->enrichDocument(
                        $resource,
                        $doc,
                        $processId
                    );
                }
                $documents[] = $doc;
            } catch (Throwable $e) {
                $this->indexerProgressHandler->error($e);
            }
```

This ensures that the indexer runs stably even if an unexpected problem occurs in individual cases. The errors are logged.

sitepark-schaeper replied:
Yes, that is what I mean.
It would have to be on the interface and every implementation that could throw.

Catching everything should be fine.

sitepark-veltrup replied:
added this

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/PersonDocumentEnricher.php:
```diff
@@ -0,0 +1,168 @@
+            if ($productLocation === null) {
+                continue;
+            }
+            $productResource = $this->resourceLoader->load($productLocation);
```
this might also fail

sitepark-veltrup replied:
see above

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/ProductDocumentEnricher.php:
```diff
@@ -0,0 +1,107 @@
+                continue;
+            }
+
+            $primaryOrganisationResource = $this->resourceLoader->load(
+                $primaryOrganisationLocation
+            );
```
another point of possible failures

sitepark-veltrup replied:
see above

sitepark-schaeper commented on src/AtooloCityGovBundle.php:
```diff
@@ -0,0 +1,34 @@
+{
+    public function build(ContainerBuilder $container): void
+    {
+        $configDir = __DIR__ . '/Resources/config';
```
We've previously named this directory just `./config` (next to `src`), why does this differ? And why are we capitalizing the `Resources`?

sitepark-veltrup replied:
Many Symfony bundles I have seen use this. I did a bit of research and it seems to still come from [Symfony 2.x}(https://symfony.com/doc/2.x/bundles/best_practices.html#directory-structure). In all subsequent Symfony versions, the structure [so](https://symfony.com/doc/7.1/bundles/best_practices.html#directory-structure) is recommended. I have changed that.

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/ContactPointDocumentEnricher.php:
```diff
@@ -0,0 +1,107 @@
+        $emailList = [];
+        $emailListData = $contactPoint['contactData']['emailList'] ?? [];
+        foreach ($emailListData as $emailData) {
+            $emailList[] = $emailData['email'];
```
Same here - What if `$emailData` does not have the array key `'email'`?

sitepark-veltrup replied:
is now being checked

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/ContactPointDocumentEnricher.php:
```diff
@@ -0,0 +1,107 @@
+        $phoneList = [];
+        $phoneListData = $contactPoint['contactData']['phoneList'] ?? [];
+        foreach ($phoneListData as $phoneData) {
+            $phoneList[] = $phoneData['phone']['nationalNumber'];
```
What if `$phoneData` does not define `['phone']['nationalNumber']`?

sitepark-veltrup replied:
is now being checked

sitepark-schaeper commented on src/Service/Indexer/Enricher/SiteKitSchema2x/ContactPointDocumentEnricher.php:
```diff
@@ -0,0 +1,107 @@
+     */
+    private function toAddressSearchValue(array $address): string
+    {
+        return trim(($address['buildingName'] ?? '') . ' '
+            .  ($address['street'] ?? '') . ' '
+            .  ($address['housenumber'] ?? ''));
```
this may produce strings like `"<buildingName>  <housenumber>"` (with two spaces) or `" 
<street> <housenumber>"` (with a leading space) or even `"  "`. Does this matter? If not, then why are we calling `trim` and why only on the `'buildingName'`)?

sitepark-veltrup replied:
The compound string is not output but used as an index value, in which the individual words are split up anyway. It therefore does not matter whether several spaces are contained one after the other.

sitepark-schaeper reviewed (commented):
First of all I cannot really comment on the contents of the `DocumentEnricher`s.  

Then I believe for our github CI it's still required that `./var/cache`, `./var/log` exist, right? Otherwise we should remove those and add this to the `composer.json`:
```json
    "archive": {
        "exclude": [
            "/*",
            "/.*",
            "!/composer.json",
            "!/src"
        ]
    }
```
We would not have to take care of all the cache/output directories for our tooling and have separate composer `scripts` to generate reports.
Also `phive` should then copy instead of link it's phars (which seems to be inconsistent right now; may be worth looking into).  

And lastly the tests are missing `declare(strict_types=1);`.

sitepark-veltrup commented:
> Then I believe for our github CI it's still required that `./var/cache`, `./var/log` exist, right? Otherwise we should remove those and add this to the `composer.json`:
> 
> ```json
>     "archive": {
>         "exclude": [
>             "/*",
>             "/.*",
>             "!/composer.json",
>             "!/src"
>         ]
>     }
> ```
> 
> We would not have to take care of all the cache/output directories for our tooling and have separate composer `scripts` to generate reports. Also `phive` should then copy instead of link it's phars (which seems to be inconsistent right now; may be worth looking into).
> 

This all currently depends on the github actions and is currently required. We may still have to improve the github actions. But that's a different task.

I have added `Archive

> And lastly the tests are missing `declare(strict_types=1);`.

added

sitepark-schaeper commented on test/Service/Indexer/Enricher/SiteKitSchema2x/OrganisationDocumentEnricherTest.php:
```diff
@@ -1,13 +1,12 @@
 <?php
 
-declare(strict_types=1);
-
```
Why remove this?

sitepark-veltrup replied:
ups, I guess too many undo's
